### ISSUE TITLE
Object Detection using Keras

### ISSUE THREAD

**Issue by Walid-Ahmed** [fchollet/keras#4869]
Can you please advice how to to do Object Detection using Keras?

**Comment by bstriner** [comment]
What problem exactly are you looking at? There seems to be a useful post about object recognition here:

http://machinelearningmastery.com/object-recognition-convolutional-neural-networks-keras-deep-learning-library/

Are you trying to get bounding boxes, multiple labels, or just single labels? What is your training set?

Cheers,
Ben

**Comment by Walid-Ahmed** [comment]
@bstriner 
I need to get bounding boxes for different objects from different classes in a single image.

**Comment by santoshgsk** [comment]
@Walid-Ahmed you got any resource?

**Comment by Walid-Ahmed** [comment]
@santoshgsk 
I got a dataset of 20,000 images.

**Comment by bstriner** [comment]
@Walid-Ahmed you'll want to start with looking around sites like arxiv to see what some of the latest models are. If you find something specific you want to implement someone can help you.

Some interesting ideas:
https://arxiv.org/pdf/1312.2249v1.pdf
https://pdfs.semanticscholar.org/713f/73ce5c3013d9fb796c21b981dc6629af0bd5.pdf

Predicting multiple boxes normally involves some post-processing to count the boxes from the neural network activations. A relatively simple approach is to just train a CNN to predict pixel or superpixel labels, and draw a box around the pixels. Another simple approach is to train a sliding window at multiple scales. You then select local maxima, or some other subset of windows.

Multiclass multiple boxes can get trickier. In one example, you train a network to not only produce masks, but label left, right, up and down. This helps disambiguate touching objects.

Cheers

**Comment by MartinThoma** [comment]
The [Overfeat paper](https://arxiv.org/abs/1312.6229) is certainly interesting (I hope somebody wants to [write a summary](http://www.shortscience.org/paper?bibtexKey=journals/corr/1312.6229)?)

**Comment by iamsiva11** [comment]
@Walid-Ahmed Any luck with the bounding box approach?

**Comment by achainan** [comment]
Take a look at the below. Probably what you are looking for.
https://github.com/yhenon/keras-frcnn

**Comment by Walid-Ahmed** [comment]
@iamsiva11 
I am  now using the code from
https://github.com/rykov8/ssd_keras

**Comment by gauravk97** [comment]
How does one go about handling variable number of objects per image for this problem of object detection?

**Comment by Walid-Ahmed** [comment]
The region proposal network used in object detection has a fixed number of candidate bounding boxes in which objects can be found so the network structure can still be fixed.

**Comment by Abhijit-2592** [comment]
Tensorflow's object detection API is the best resource available online to do object detection. I have a small [blog post](https://abhijit-2592.github.io/Keras-with-TFODAPI/ ) that explains how to integrate Keras with the object detection API, with this small trick you will be able to convert any classification model trained in Keras to an object detection model using the API.

**Comment by samrawitergrtr** [comment]
Can you please advice how to to do single label Object Detection and localization  using Keras?
